posthog/queries/retention.py

```diff
@@ -1,62 +1,203 @@
                 ],
-                "label": "{} {}".format(period, first_day),
-                "date": (date_from + _determineTimedelta(first_day, period)[0]),
+                "label": "{} {}".format(filter.period, first_day),
+                "date": (date_from + self.determineTimedelta(first_day, filter.period)[0]),
             }
             for first_day in range(total_intervals)
         ]
 
         return result
 
+    def _execute_sql(
+        self,
+        filter: Filter,
+        date_from: datetime.datetime,
+        date_to: datetime.datetime,
+        target_entity: Entity,
+        team: Team,
+    ) -> Dict[Tuple[int, int], Dict[str, Any]]:
+
+        period = filter.period
+        events: QuerySet = QuerySet()
+
+        if target_entity.type == TREND_FILTER_TYPE_EVENTS:
+            events = Event.objects.filter_by_event_with_people(event=target_entity.id, team_id=team.id)
+        elif target_entity.type == TREND_FILTER_TYPE_ACTIONS:
+            events = Event.objects.filter(action__pk=target_entity.id).add_person_id(team.id)
+
+        filtered_events = events.filter(filter.date_filter_Q).filter(filter.properties_to_Q(team_id=team.pk))
+        trunc, fields = self._get_trunc_func("timestamp", period)
+        first_date = filtered_events.annotate(first_date=trunc).values("first_date", "person_id").distinct()
+
+        event_query, events_query_params = filtered_events.query.sql_with_params()
+        reference_event_query, first_date_params = first_date.query.sql_with_params()
+
+        final_query = """
+            SELECT
+                {fields}
+                COUNT(DISTINCT "events"."person_id"),
+                array_agg(DISTINCT "events"."person_id") as people
+            FROM ({event_query}) events
+            LEFT JOIN ({reference_event_query}) first_event_date
+              ON (events.person_id = first_event_date.person_id)
+            WHERE timestamp >= first_date
+            GROUP BY date, first_date
+        """.format(
+            event_query=event_query, reference_event_query=reference_event_query, fields=fields
+        )
+
+        start_params = (date_from, date_from) if period == "Month" or period == "Hour" else (filter.date_from,)
+
+        with connection.cursor() as cursor:
+            cursor.execute(
+                final_query, start_params + events_query_params + first_date_params,
+            )
+            data = namedtuplefetchall(cursor)
+
+            scores: dict = {}
+            for datum in data:
+                key = round(datum.first_date, 1)
+                if not scores.get(key, None):
+                    scores.update({key: {}})
+                for person in datum.people:
+                    if not scores[key].get(person, None):
+                        scores[key].update({person: 1})
+                    else:
+                        scores[key][person] += 1
+
+        by_dates = {}
+        for row in data:
```

Comment:
ch_retention doesn't have any of this yet which is a [logged issue ](https://github.com/PostHog/posthog/issues/2280) now